common/unsafe/src/main/java/org/apache/spark/unsafe/memory/ByteArrayMemoryBlock.java

```diff
@@ -0,0 +1,190 @@
+    return new ByteArrayMemoryBlock(array, offset, size);
+  }
+
+  public byte[] getByteArray() { return array; }
+
+  /**
+   * Creates a memory block pointing to the memory used by the byte array.
+   */
+  public static ByteArrayMemoryBlock fromArray(final byte[] array) {
+    return new ByteArrayMemoryBlock(array, Platform.BYTE_ARRAY_OFFSET, array.length);
+  }
+
+
+  public final int getInt(long offset) {
+    // UTF8String.getPrefix() assumes data is 4-byte aligned
+    assert(offset + 4 - Platform.BYTE_ARRAY_OFFSET <= ((array.length + 3) / 4) * 4);
+    return Platform.getInt(array, offset);
+  }
+
+  public final void putInt(long offset, int value) {
+    assert(offset + 4 - Platform.BYTE_ARRAY_OFFSET <= array.length);
+    Platform.putInt(array, offset, value);
+  }
+
+  public final boolean getBoolean(long offset) {
+    assert(offset + 1 - Platform.BYTE_ARRAY_OFFSET <= array.length);
+    return Platform.getBoolean(array, offset);
+  }
+
+  public final void putBoolean(long offset, boolean value) {
+    assert(offset + 1 - Platform.BYTE_ARRAY_OFFSET <= array.length);
+    Platform.putBoolean(array, offset, value);
+  }
+
+  public final byte getByte(long offset) {
+    return array[(int)(offset - Platform.BYTE_ARRAY_OFFSET)];
+  }
+
+  public final void putByte(long offset, byte value) {
+    array[(int)(offset - Platform.BYTE_ARRAY_OFFSET)] = value;
+  }
+
+  public final short getShort(long offset) {
+    assert(offset + 2 - Platform.BYTE_ARRAY_OFFSET <= array.length);
+    return Platform.getShort(array, offset);
+  }
+
+  public final void putShort(long offset, short value) {
+    assert(offset + 2 - Platform.BYTE_ARRAY_OFFSET <= array.length);
+    Platform.putShort(array, offset, value);
+  }
+
+  public final long getLong(long offset) {
+    // UTF8String.getPrefix() assumes data is 8-byte aligned
+    assert(offset + 8 - Platform.BYTE_ARRAY_OFFSET <= ((array.length + 7) / 8) * 8);
+    return Platform.getLong(array, offset);
+  }
+
+  public final void putLong(long offset, long value) {
+    assert(offset + 8 - Platform.BYTE_ARRAY_OFFSET <= array.length);
+    Platform.putLong(array, offset, value);
+  }
+
+  public final float getFloat(long offset) {
+    assert(offset + 4 - Platform.BYTE_ARRAY_OFFSET <= array.length);
+    return Platform.getFloat(array, offset);
+  }
+
+  public final void putFloat(long offset, float value) {
+    assert(offset + 4 - Platform.BYTE_ARRAY_OFFSET <= array.length);
+    Platform.putFloat(array, offset, value);
+  }
+
+  public final double getDouble(long offset) {
+    assert(offset + 8 - Platform.BYTE_ARRAY_OFFSET <= array.length);
+    return Platform.getDouble(array, offset);
+  }
+
+  public final void putDouble(long offset, double value) {
+    assert(offset + 8 - Platform.BYTE_ARRAY_OFFSET <= array.length);
+    Platform.putDouble(array, offset, value);
+  }
+
+  public final void copyFrom(byte[] src, long srcOffset, long dstOffset, long length) {
```

Comment:
let me raise my question again. Since `MemoryBlock` has `baseObj` and `baseOffset`, We can implement `getInt`, `copyTo`, etc. in `MemoryBlock` as final method. However we put them in concrete implementations, for performance concerns. This is not ideal as the implementation code are very similar, so we should only do it if it really has perf benefit.

I'd like to create a benchmark for `MemoryBlock` and keep the benchmark in the codebase, so it's easy to verify the perf gains.

Particularly, I'm a little worried about the `copyFrom`. By the theory, `Platform.copyMemory` should run faster if both the `srcObj` and `dstObj` has concrete type. To make `srcObj` has concrete type, we create many varieties of `copyFrom`, like `copyFom(byte[] ...)`, `copyFrom(long[] ...)`, etc. To make `dstObj` has concrete type, we implement various `copyFrom` in all kinds of memory blocks. My question is, do both of them have perf benefit?

Reply:
Sure, I will create and run benchmark programs this weekend. Thus, I did not change this part in the last commit. I should have mentioned this postpone. Sorry.